env.js

```diff
@@ -1,13 +1,14 @@
 var log = require('loglevel');
 
 module.exports = function() {
-  if (process.env.COUCH2PG_DEBUG) {
+  if (process.env.COUCH2PG_DEBUG || true) {
     log.setDefaultLevel('debug');
   } else {
     log.setDefaultLevel('info');
   }
 
   return {
+    legacyMode: process.env.LEGACY_MODE || false,
```

Comment:
I think doing this in a forward thinking way is really hard.

The problem with feature detection is that you never know if you have the whole story.

For example, we want some views to be created in 2.6 to make dealing with contacts easier. The detection, creation and management etc of this is now done using a DB migration tool. To do this with feature detection, we would need to build a system that checks for un-created views (if it creates two views, do you check one of them or both? What if you want to update that view or package of changes later?) and then determines if they need to be created: in this case, if there are any documents of type `contact`. You need to do this every time, because maybe the first time you sync up there are no contact documents yet. This seems really terrible compared to using a migration management tool.

Instead of feature detection you could group functionality broader than a single feature (allowing you to use things like db migration without it being a giant hassle), e.g. "xmlforms", which encompasses all work needed for that very broad feature. For example, in the future it may be required for us to extract out the form version from the XML and put that in a queryable table. Or extract out GPS coords from older forms that don't have them exposed in the JSON.

This doesn't really map well to a "feature" though, even a broad one like "xml forms". In future versions we could keep that broad feature but the code could change to not require (or even invalidate) the work that code needs to do. For example, in 3.0 we might already extract out the form version into the JSON, or maybe even have a system where the concept of form version doesn't exist anymore. 

As you said, you're not a fan of version detection, which is the next obvious step in this thought experiment, so we can skip over that one.

The route I was going with for "legacy mode" was that legacy == 0.4, not legacy == 2.x, *but only because the 1.0 of this comes out when both exist in production*. If we had built this when just 0.4 existed there would have been no legacy mode. When medic-webapp 3.0 comes along and is substantially different (or 2.8, or anything really), the idea was to simply not support 2.x in master anymore. The old releases still exist, and if there are bug fixes etc that are required we can check out the tag and go through that process. Once this changed version is a bit stable the idea would *also* be to drop legacy mode from master completely, and thus master of medic-couch2pg === master of medic-webapp, and everything else is tags.